rfd/0064-audit-modes.md

```diff
@@ -0,0 +1,190 @@
+sessions that can't be recorded or have audit logging failures.
+
+## Why
+
+In cases where Teleport is configured to record sessions on the node using the
+asynchronous format, disk failures (for example, no space left) causes the node
+to be inaccessible. Where Teleport is the only way of accessing those resources,
+administrators cannot perform the necessary actions on these nodes to bring them
+back to a normal state.
+
+## Goals
+
+* Prevent nodes from being inaccessible when there are audit errors.
+* Enable users to configure the strictness of audit failures.
+
+**Non-goals:**
+
+* Infrastructure-related metrics (such as disk space left).
+* Enable access when the recording mode is set to the proxy.
+
+## Details
+
+Audit modes will define how Teleport proceeds in case of audit failures. Those
+failures vary depending on how session recording is configured:
+
+| Recording mode                 | Failures handled                                             |
+| ------------------------------ | ------------------------------------------------------------ |
+| `node` and `proxy`             | I/O errors while writing audit logs/session recording data on the server disk. |
+| `node-async` and `proxy-async` | Connection errors with Proxy/Auth server while streaming audit logs/session recording data. |
+
+There are going to be two audit modes:
+
+* **Strict**: Teleport won't tolerate session recording or audit logging to fail.
+  In practice, if Teleport cannot record the session, it will not start or will
+  be terminated (in case it is already in progress). This mode will be the
+  recommended one as it guarantees all sessions are being recorded.
+* **Best-effort**: Audit failures are tolerated in this mode, meaning that Teleport
+  will allow sessions without recording them. System administrators might use
+  this mode to be able to access resources continually.
+
+## UX
+
+### Configuration
+
+The configuration is going to be done at the role level. The role option
+`record_session` will be extended to hold the audit mode values: `strict` and
+`best_effort`. It will have one entry per session kind (desktop, databases,
+applications, Kubernetes, and SSH) and an `all` entry for cases where users
+want the same behavior for all kinds.
+
+**Note**: Currently, the option  `record_session.desktop` uses a boolean value to
+disable/enable session recording. It will have to be converted into a new
+format: `true` will be charged to the default audit mode, and `false` will be
+`off`. Disabling session recording is out of the scope of this RFD, meaning that
+the `off` value will only be present on the `desktop` kind.
+
+**Role example (with all session kinds):**
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: alice
+spec:
+  allow:
+    ...
+  options:
+    record_session:
+      all: strict|best_effort
+      ssh: strict|best_effort
+      app: strict|best_effort
+      db: strict|best_effort
```

Comment:
We also don't really have session recording for app/db access currently (only audit logging). So we may leave this out of initial scope I think.

Reply:
This is a little bit confusing to me. Application "records sessions" use audit logging, as each request is an event, persisted to disk (in case of Teleport is configured to be async). Hence, the initial issue also affects application access. I'm not 100% sure about database access, but it also relies on the [filesessions](https://github.com/gravitational/teleport/blob/master/lib/srv/db/streamer.go#L93), which is where recordings are persisted.

Do you think we should not cover audit logging in this RFD? or leave these other "session recordings" outside its scope?